### ISSUE TITLE
[ngrx-ducks] selector decorator

### ISSUE THREAD

**Issue by skydever** [co-IT/co-it#10]
Hi there!

We were discussing the possibility of selector methods inside of a `Duck` in our team today and we came up with the following proposal:

* Create a `@Selector()` decorator to use at a `Duck` like this:
```typescript
export interface CounterState {
  counter: number;
}

@Ducksify<CounterState>({
  initialState: {
    counter: 0
  },
  featureName: 'counter'
})
export class Counter {
  @Action('[Counter] Add value')
  add(state: CounterState, payload: number): number {
    return {
      ...state
      counter: counter + payload
    };
  }

  @Selector()
  // maybe: @Selector('counter') or @Selector('object.child.value') as a utility param
  getCounterValue(state: CounterState): number {
    return state.counter;
  }
}
```

* The `Duck` ducksified class then should contain the following method for a usage like this:
```typescript
  counter$: Observable<number>;
  constructor(@Inject(Counter) private counter: Duck<Counter>) {
    this.counter$ = this.counter.getCounterValue();
  }
```

Internally the `getCounterValue(stateSlice)` method is a wrapper for this:
```typescript
  visitFeature = createFeatureSelector<CounterState>(
    featureName
  ); // from @Ducksify({featureName: string;})

  getCounterValue(): Observable<number> {
    return this.store.pipe(select(
      createSelector(visitFeature, s => this.originalGetCounterValueMethod(s))
    ));
  }
```

We think this would be awesome, the facade concept is totally fullfilled - you have methods to set and get values from the store handled by the `Duck`. What do you think about it? @GregOnNet 

ps: sorry if the schematics are delayed a bit, we are totally concentrated on the `EntityDuck` and the `@Selector()` possibilities at the moment.

**Comment by skydever** [comment]
... and some additional thoughts:

* the schematics can create a `NgModule` that does the setup for the duck:

```typescript
@NgModule({
  imports: [
    StoreModule.forFeature(
      /**
       *  a static method to get the reducer + featureName *out of the box*, 
       *  or a wrapper around the
       *  function counterReducer(state, action): CounterState { 
       *    return reducerFrom(Counter)(state, action);
       *  }
       */
      Duck.featureName(Counter),
      Duck.reducer(Counter)
    ),
    EffectsModule.forFeature([CounterEffects])
  ]
})
export class CounterModule{}
```

* afterward you just import `CounterModule` inside the eg. `AppModule`, or `CoreModule`:

```typescript
@NgModule({
  imports: [
    ...
    CounterModule
  ]
  ...
})
export class AppModule {}
```

... or is this possible?
```typescript
  Duck.module(Counter)
```

**Comment by GregOnNet** [comment]
## DucksModule

I tried to get this to run in various ways, but due AoT restrictions I failed.
The problem is that we need to call the function `ducksify` to transform the class into its service representation.
Unfortunately AoT cannot resolve the process passing types to a function that loops through it to provide a service.

The `@Ducksify` decorator is the only way I managed to get it to work

## Selector

Honestly, I did not think about to introduce another Decorator `@Select`. :)
In the first step, I wanted to do something like...

```ts
@Ducksify({ /* ... */ })
export class Counter {
  private counter = createFeatureSelector('counter');
  readonly current = createSelector(counter, counter => counter.count)

  // ...
}
```

... My intention is to reuse the API of ngrx-selectors that the community is already fimiliar with.
**BUT** maybe it is not a good practise to place selectors inside the class since you wont be able to combine selectors with each other. 🤔

**Comment by skydever** [comment]
### DucksModule
I see, thank you! Maybe the AoT compilation will change in the future to make such things possible :wink:

### Selector
We have a lot of use cases where a simple `@Select()` decorator like described above would help a lot to reduce lines of code to be maintained.

At the moment we create the selectors plain **ngrx** way, but we add them as a static property to the duck:

```typescript
export class CounterDuck {
  static featureName = 'counter';
  static reducer = reducer; // links to the reducer method at the end of a duck file
  static select = fromCounter; // is an object containing all selectors
```

When we setup the `Duck` in a feature module it looks like this:
```typescript
@NgModule({
  imports: [
    StoreModule.forFeature(Counter.featureName, Counter.reducer),
    EffectsModule.forFeature([CounterEffects])
  ]
```

When we read things from the `Counter` store:
```typescript
this.counter.pick(Counter.select.current)
```

This way we have code completion for existing selectors, but it requires everybody at the team to stick to a convention. Schematics help a lot, but as you mention in the [quick start guide](https://github.com/co-IT/co-it/blob/master/ngrx/ducks/docs/quick-start.md) having less code generated that needs to be maintaind would be great.

Maybe it is too special, I am not sure. I really like the idea of the `@Select` decorator somehow :smile:

**Comment by GregOnNet** [comment]
Hi,

it is not an edge case.
Adding the selectors was part of my agenda.

It already landed in version 7.4.0 which is now available to you.
You are now able to add selectors directly to the duck.

For now, I do not implement `@Selector` since it adds more abstraction and restricts possibilities to combine multiple selectors.

Please refer to the following sample and let me know what you think.
([Demo ⚡️](https://stackblitz.com/edit/ngrx-ducks?file=src%2Fapp%2Fstore%2Fcounter%2Fcounter.selectors.ts))

## Selectors

```ts
import { createFeatureSelector, createSelector } from '@ngrx/store';
import { CounterState } from './counter.state';

const visitCounter = createFeatureSelector<CounterState>('counter');

export const currentCount = createSelector(
  visitCounter,
  c => c.count
);
export const isLoading = createSelector(
  visitCounter,
  c => c.isLoading
);
```

## Duck

```ts
import { Action, effect, InitialState, Ducksify, DucksifiedAction, reducerFrom } from '@co-it/ngrx-ducks';
import { CounterState } from './counter.state';

import * as selectors from './counter.selectors';

@Ducksify<CounterState>({
  initialState: {
    count: 0,
    isLoading: true
  }
})
export class Counter {
  readonly set = effect<number>('[Counter] Load Count');

  current$ = selectors.currentCount;
  isLoading$ = selectors.isLoading;

  @Action('[Counter] Increment value')
  increment(state: CounterState, payload: number): CounterState {
    return {
      ...state,
      count: state.count + payload
    };
  }

  @Action('[Counter] Decrement value')
  decrement(state: CounterState, payload: number): CounterState {
    return {
      ...state,
      count: state.count - payload
    };
  }

  @Action('[Counter] Set value')
  override(state: CounterState, payload: number): CounterState {
    return {
      ...state,
      count: payload,
      isLoading: false
    };
  }
}
```

## Usage
```ts
import { Component, Inject } from '@angular/core';
import { Duck } from '@co-it/ngrx-ducks';
import { Observable } from 'rxjs';
import * as fromCounter from '../store/counter';
import { Counter } from '../store/counter';

@Component({
  selector: 'counter',
  templateUrl: './counter.component.html',
  styleUrls: ['./counter.component.scss']
})
export class CounterComponent {
  count$: Observable<number>;
  isLoading$: Observable<boolean>;
  
  constructor(@Inject(Counter) private counter: Duck<Counter>) {
    this.counter.set.dispatch(10);

    this.count$ = this.counter.current$;
    this.isLoading$ = this.counter.isLoading$;

    /** Furthermore, you can pass any selector you want using the pick-selector
     * 
     * this.isLoading$ = this.counter.pick(fromCounter.isLoading);
     * this.count$ = this.counter.pick(fromCounter.currentCount);
     * 
     */
  }

  increment() {
    this.counter.increment(1);
  }

  decrement() {
    this.counter.decrement(1);
  }
}
```

**Comment by skydever** [comment]
Hi!

### Selectors

This is an amazing new feature!! 🎉 The `Duck` is a 💯 facade now! 👍 

`@Selector()` could be a helper for simple selectors ... but the more I think about it, it is not that less code and as you mentioned the community is already familiar with the **ngrx** selector API, and drifting not too far from it is good I guess. The logic is also splitted into multiple files, I like that. Our intention was mainly to get code completion at the usage of a `Duck`, and this is possible now.

### Module setup

I had an idea for the `reducerFrom` method:

```typescript
export function duckReducerFrom(duckClass) {
  return (state, action) => reducerFrom(duckClass)(state, action);
}
```

then you can do things like that

```typescript
StoreModule.forFeature('counter', duckReducerFrom(Counter))
```

This acutally does work [here](https://stackblitz.com/edit/ngrx-ducks-reducer?file=src%2Fapp%2Fstore%2Fduck-reducer-from.ts) ⚡️, I am importing it at the `CounterModule` to try that (the `ActionReducerMap` also does work). No need to create the reducer method for every duck anymore 🎉 Stackblitz is using AoT, right?

**Comment by GregOnNet** [comment]
As far as I know, StackBlitz works with JIT-Compiler.
I downloaded your sample and started the production build.

```
npm run build -- --prod
```

The compilation fails.
I am very familiar with the error shown below. 🙄🤣 

![image](https://user-images.githubusercontent.com/444278/53290537-b6f2ae80-37a5-11e9-96db-8a538f18c202.png)

**Comment by GregOnNet** [comment]
I close this issue since it is related to the selector feature that has been released in version 7.4.0 🎉

**Comment by skydever** [comment]
... would have been too easy 😢 I can imagine that you already tried that 😄 Thx for the info about stackblitz and jit. The selector feature is great 👍
